File: src/symmetry/symmetry_elements/ScrewAxis.jl
#   Copyright 2025 Velexi Corporation
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

"""
ScrewAxis type and functions
"""
# --- Exports

# Types
export ScrewAxis

# Constants
export a_2_1, a_4_1, a_4_2, a_4_3
export b_2_1, b_4_1, b_4_2, b_4_3
export c_2_1, c_4_1, c_4_2, c_4_3
export c_3_1, c_3_2
export c_6_1, c_6_2, c_6_3, c_6_4, c_6_5

# --- Types

"""
    ScrewAxis

Type representing a screw axis

Supertype: [`SymmetryElement`](@ref)
"""
struct ScrewAxis <: SymmetryElement
    # --- Fields

    # rotation order
    n::Int

    # number of translation steps of size 1/n following rotation by 2π/n
    m::Int

    # direction of rotation axis in unit cell fractional coordinates
    direction::Tuple{Rational,Rational,Rational}

    # a point on the rotation axis in unit cell fractional coordinates
    location::Tuple{Rational,Rational,Rational}

    # --- Constructor

    function ScrewAxis(
        n::Int,
        m::Int,
        direction::Tuple{<:Real,<:Real,<:Real},
        location::Tuple{<:Real,<:Real,<:Real},
    )

        # Enforce constraints
        if n ≤ 0
            throw(ArgumentError("`n` must be positive (n=$n)"))
        end

        if m ≤ 0
            throw(ArgumentError("`m` must be positive (m=$m)"))
        end

        if m ≥ n
            throw(ArgumentError("`m` must be less than `n` (n=$n,m=$m)"))
        end

        # Return new ScrewAxis
        return new(n, m, direction, location)
    end
end

# Outer constructor
"""
    ScrewAxis(
        n::Int,
        m::Int,
        direction::Tuple{<:Real,<:Real,<:Real};
        location::Tuple{<:Real,<:Real,<:Real}=(0, 0, 0)
    )

Construct a ScrewAxis object that

* has an axis by `direction` and `location`

and

* couples a rotation by `2π/n` and a translation along the rotation axis by `m/n` (in unit
  cell fractional coordinates).

Keyword Arguments
=================
- `location`: a point on the rotation axis in unit cell fractional coordinates
"""
function ScrewAxis(
    n::Int,
    m::Int,
    direction::Tuple{<:Real,<:Real,<:Real};
    location::Tuple{<:Real,<:Real,<:Real}=(0, 0, 0),
)
    return ScrewAxis(n, m, direction, location)
end

# --- Functions/Methods

import Base.:(==)
import LinearAlgebra: dot

function Base.:(==)(x::ScrewAxis, y::ScrewAxis)
    # Check that rotation orders and translation steps are equal
    if x.n != y.n
        return false
    end

    if x.m != y.m
        return false
    end

    # Check that directions are the same
    if dot(x.direction, y.direction)^2 !=
        dot(x.direction, x.direction) * dot(y.direction, y.direction)
        return false
    end

    # Check that line through both locations is in the same direction as both lines
    delta = (x.location[i] - y.location[i] for i in 1:3)
    return dot(delta, x.direction)^2 == dot(delta, delta) * dot(x.direction, x.direction)
end

# --- Constants

const a_2_1 = ScrewAxis(2, 1, (1, 0, 0))
const a_4_1 = ScrewAxis(4, 1, (1, 0, 0))
const a_4_2 = ScrewAxis(4, 2, (1, 0, 0))
const a_4_3 = ScrewAxis(4, 3, (1, 0, 0))

const b_2_1 = ScrewAxis(2, 1, (0, 1, 0))
const b_4_1 = ScrewAxis(4, 1, (0, 1, 0))
const b_4_2 = ScrewAxis(4, 2, (0, 1, 0))
const b_4_3 = ScrewAxis(4, 3, (0, 1, 0))

const c_2_1 = ScrewAxis(2, 1, (0, 0, 1))
const c_3_1 = ScrewAxis(3, 1, (0, 0, 1))
const c_3_2 = ScrewAxis(3, 2, (0, 0, 1))
const c_4_1 = ScrewAxis(4, 1, (0, 0, 1))
const c_4_2 = ScrewAxis(4, 2, (0, 0, 1))
const c_4_3 = ScrewAxis(4, 3, (0, 0, 1))
const c_6_1 = ScrewAxis(6, 1, (0, 0, 1))
const c_6_2 = ScrewAxis(6, 2, (0, 0, 1))
const c_6_3 = ScrewAxis(6, 3, (0, 0, 1))
const c_6_4 = ScrewAxis(6, 4, (0, 0, 1))
const c_6_5 = ScrewAxis(6, 5, (0, 0, 1))

File: pysrc/xtallography/symmetry/symmetry_elements/screw_axis.py
#   Copyright 2025 Velexi Corporation
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
"""
ScrewAxis class
"""

# --- Imports

# Standard library
from dataclasses import dataclass
from fractions import Fraction
from typing import Optional

# External packages
import numpy

# Local packages/modules
from .symmetry_element import SymmetryElement
from ... import _JL


# --- Classes


@dataclass(frozen=True)
class ScrewAxis(SymmetryElement):
    """
    Class representing a screw axis

    Fields
    ------
    * `n` (int): rotation order

    * `m` (int): number of translation steps of size 1/n following rotation by 2π/n

    * `direction` (tuple): direction of rotation axis

    * `location` (tuple): a point on the rotation axis in unit cell fractional coordinates
    """

    # --- Fields

    n: int
    m: int
    direction: tuple
    location: tuple

    # --- Initializer

    def __init__(
        self, n: int, m: int, direction: tuple, location: Optional[tuple] = None
    ):
        """
        Initialize ScrewAxis object.

        Parameters
        ----------
        `n`: rotation order

        `m`: number of translation steps of size 1/n following rotation by 2π/n

        `direction`: direction of rotation axis

        `location`: a point on the rotation axis in unit cell fractional coordinates
        """
        # --- Check arguments

        if n <= 0:
            raise ValueError(f"`n` must be positive (n={n})")

        if m <= 0:
            raise ValueError(f"`m` must be positive (m={m})")

        if m >= n:
            raise ValueError(f"`m` must be less than `n` (n={n},m={m})")

        # check that `direction` is a 3-tuple
        if len(direction) != 3:
            raise ValueError(f"`direction` must be a 3-tuple (direction={direction})")

        # convert `direction` to Fraction objects
        try:
            direction = tuple(Fraction(x) for x in direction)
        except ValueError as error:
            raise ValueError(
                "all elements of `direction` must convertible to Fraction objects "
                f"(direction={direction}). "
                f"[caused by {type(error)}({error})]"
            )

        # set default value for `location`
        if location is None:
            location = (Fraction(0), Fraction(0), Fraction(0))

        else:
            # check that `location` is a 3-tuple
            if len(location) != 3:
                raise ValueError(f"`location` must be a 3-tuple (location={location})")

            # convert `location` to Fraction objects
            try:
                location = tuple(Fraction(x) for x in location)
            except ValueError as error:
                raise ValueError(
                    "all elements of `location` must convertible to Fraction objects "
                    f"(location={location}). "
                    f"[caused by {type(error)}({error})]"
                )

        # --- Initialize field values

        # for frozen DataClasses, field values cannot be set directly
        object.__setattr__(self, "n", n)
        object.__setattr__(self, "m", m)
        object.__setattr__(self, "direction", direction)
        object.__setattr__(self, "location", location)

    # --- Methods

    def to_julia(self):
        """
        Convert Python ScrewAxis object to a Julia ScrewAxis object.
        """
        return _JL.ScrewAxis(self.n, self.m, self.direction, self.location)

    @classmethod
    def from_julia(cls, screw_axis_jl: _JL.ScrewAxis):
        """
        Convert a Julia ScrewAxis object to a Python ScrewAxis object.

        Parameters
        ----------
        screw_axis_jl: Julia ScrewAxis object

        Return value
        ------------
        Python ScrewAxis object
        """
        # Check arguments
        if not _JL.isa(screw_axis_jl, _JL.ScrewAxis):
            raise ValueError(
                "`screw_axis_jl` must be a Julia `ScrewAxis` object. "
                f"(screw_axis_jl={screw_axis_jl})."
            )

        # Convert screw_axis_jl to a ScrewAxis object
        return ScrewAxis(
            screw_axis_jl.n,
            screw_axis_jl.m,
            screw_axis_jl.direction,
            screw_axis_jl.location,
        )

    def __repr__(self):
        """
        Return string representation of ScrewAxis.
        """
        return (
            f"ScrewAxis(n={self.n},m={self.m},direction={self.direction},"
            f"location={self.location})"
        )

    def __eq__(self, other):
        """
        Return True if `self` and `other`are identical symmetry elements; otherwise,
        return False.

        Parameters
        ----------
        other:  SymmetryElement object to compare against
        """
        if not isinstance(other, type(self)):
            return False

        # Check that orders are equal
        if self.n != other.n:
            return False

        # Check that the number of translation steps are equal
        if self.m != other.m:
            return False

        # Check that directions are the same
        if numpy.dot(self.direction, other.direction) ** 2 != numpy.dot(
            self.direction, self.direction
        ) * numpy.dot(other.direction, other.direction):
            return False

        # Check that line through both locations is in the same direction as both lines
        delta = tuple(self.location[i] - other.location[i] for i in range(3))
        return numpy.dot(delta, self.direction) ** 2 == numpy.dot(
            delta, delta
        ) * numpy.dot(self.direction, self.direction)
